File: src/services/anime/find.ts
import { searchAnime } from './search'
import { findAnimeByDescription as findAnimeByDescriptionOpenAI } from '@/services/openai/completions'
import { findAnimeByDescriptionClaude } from '@/services/claude/completions'
import { IS_OPENAI_ENABLED, IS_CLAUDE_ENABLED, OPENAI_API_KEY, CLAUDE_API_KEY } from '@/config/constants'
import type { AnimeRecommendation } from '@/types/openai'
import type { AnimeMatch } from '@/types/anilist'

/**
 * Find anime using AI description and return AniList details
 */
export async function findAnimeWithDetails(description: string): Promise<AnimeMatch[]> {
  if (!IS_OPENAI_ENABLED && !IS_CLAUDE_ENABLED) {
    throw new Error('AI is not configured. Please set OPENAI_API_KEY or CLAUDE_API_KEY environment variable to use AI-powered anime search.')
  }

  try {
    // Prefer OpenAI if both keys are present
  let recommendations: AnimeRecommendation[] = []
    if (IS_OPENAI_ENABLED && OPENAI_API_KEY) {
      recommendations = await findAnimeByDescriptionOpenAI(description)
    } else if (IS_CLAUDE_ENABLED && CLAUDE_API_KEY) {
      recommendations = await findAnimeByDescriptionClaude(description)
    }

    const matches: AnimeMatch[] = []

    // Search for each recommendation on AniList
    for (const rec of recommendations) {
      try {
        const searchResults = await searchAnime(rec.title, 1, 5)
        if (searchResults.media.length > 0) {
          // Use the first (most relevant) result
          const anime = searchResults.media[0]
          matches.push({
            anime,
            reason: rec.reason,
            confidence: rec.confidence
          })
        }
      } catch (error) {
        console.warn(`Could not find anime "${rec.title}" on AniList:`, error)
        // Continue with other recommendations
      }
    }

    // Sort by confidence score
    return matches.sort((a, b) => b.confidence - a.confidence)

  } catch (error) {
    console.error('Error finding anime with details:', error)
    throw new Error('Failed to find anime based on description')
  }
}

File: go/internal/services/anilist/find.go
package anilist

import (
	"encoding/json"
	"fmt"
	"log"

	"discord-anime-bot/internal/config"
	"discord-anime-bot/internal/services/claude"
	"discord-anime-bot/internal/services/openai"
	"discord-anime-bot/internal/types"
)

// FindAnimeWithDetails finds anime using AI description and returns AniList details
func FindAnimeWithDetails(description string, cfg *config.Config) ([]types.AnimeMatch, error) {
	if !cfg.IsAIEnabled {
		return nil, fmt.Errorf("AI is not configured. Please set OPENAI_API_KEY or CLAUDE_API_KEY environment variable to use AI-powered anime search")
	}

	var recommendations []types.OpenAIRecommendation
	var err error

	if cfg.IsOpenAIEnabled {
		recommendations, err = openai.FindAnimeByDescription(description, cfg.OpenAIAPIKey)
		if err != nil {
			return nil, err
		}
	} else if cfg.IsClaudeEnabled {
		claudeClient := claude.NewClaudeClient()
		if claudeClient == nil {
			return nil, fmt.Errorf("claude is not configured. Please set CLAUDE_API_KEY environment variable to use AI-powered anime search")
		}
		jsonStr, err := claudeClient.FindAnimeByDescription(description)
		if err != nil {
			return nil, err
		}
		if err := json.Unmarshal([]byte(jsonStr), &recommendations); err != nil {
			return nil, fmt.Errorf("failed to parse Claude response: %w", err)
		}
	}

	var matches []types.AnimeMatch

	// Search for each recommendation on AniList
	for _, rec := range recommendations {
		searchResults, err := SearchAnime(rec.Title, 1, 5)
		if err != nil {
			log.Printf("Could not search for anime %q: %v", rec.Title, err)
			continue
		}

		if len(searchResults.Data.Page.Media) > 0 {
			// Use the first (most relevant) result
			anime := searchResults.Data.Page.Media[0]
			matches = append(matches, types.AnimeMatch{
				Anime:      anime,
				Reason:     rec.Reason,
				Confidence: rec.Confidence,
			})
		}
	}

	// Sort by confidence score (higher first)
	for i := 0; i < len(matches)-1; i++ {
		for j := i + 1; j < len(matches); j++ {
			if matches[i].Confidence < matches[j].Confidence {
				matches[i], matches[j] = matches[j], matches[i]
			}
		}
	}

	return matches, nil
}
